common/src/main/java/com/sequenceiq/cloudbreak/health/HikariStateHealthIndicator.java

```diff
@@ -0,0 +1,74 @@
+package com.sequenceiq.cloudbreak.health;
+
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+import javax.inject.Inject;
+import javax.sql.DataSource;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.beans.factory.annotation.Value;
+import org.springframework.boot.actuate.health.Health;
+import org.springframework.boot.actuate.health.HealthIndicator;
+import org.springframework.stereotype.Component;
+
+import com.sequenceiq.cloudbreak.util.ThreadUtil;
+import com.zaxxer.hikari.HikariDataSource;
+
+@Component
+public class HikariStateHealthIndicator implements HealthIndicator {
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(HikariStateHealthIndicator.class);
+
+    @Value("${hikari.pool.readiness.down-if-pool-is-full}")
+    private boolean setReadinessProbeDownIfPoolIsFull;
+
+    @Inject
+    private List<DataSource> dataSources;
+
+    @Inject
+    private HikariDataSourcePoolMetadataExtractor hikariDataSourcePoolMetadataExtractor;
+
+    @Override
+    public Health health() {
+        boolean poolIsFull = false;
+        Map<String, Object> details = new HashMap<>();
+        try {
+            for (DataSource dataSource : dataSources) {
+                if (dataSource instanceof HikariDataSource) {
+                    HikariDataSource hikariDataSource = (HikariDataSource) dataSource;
+                    String poolName = hikariDataSource.getPoolName();
+                    DataSourcePoolMetadata poolMetadata = hikariDataSourcePoolMetadataExtractor.extract(hikariDataSource);
+                    details.put(poolName, poolMetadata);
+                    if (poolMetadata.getActive() >= poolMetadata.getMax()) {
+                        LOGGER.warn("Application has run out of open db connections. name: {}, details: {}", poolName, poolMetadata);
+                        poolIsFull = true;
+                    }
+                }
+            }
+            LOGGER.debug("HikariPool details: {}", details);
+            threadDumpIfNeeded(poolIsFull);
+        } catch (RuntimeException exception) {
+            LOGGER.info("Failed while checking the health of hikari pool!", exception);
+        }
+        return buildHealthResponse(poolIsFull, details);
+    }
```

Comment:
I would suggest a bit of refactor to make it more readable:
```
    public Health health() {
        try {
            Map<String, DataSourcePoolMetadata> details = dataSources.stream()
                    .filter(ds -> ds instanceof HikariDataSource)
                    .map(ds -> (HikariDataSource) ds)
                    .collect(Collectors.toMap(HikariConfig::getPoolName, ds -> hikariDataSourcePoolMetadataExtractor.extract(ds)));
            Map<String, DataSourcePoolMetadata> poolsAtLimit = details.entrySet().stream()
                    .filter(ds -> ds.getValue().getActive() >= ds.getValue().getMax())
                    .collect(Collectors.toMap(Map.Entry::getKey, Map.Entry::getValue));
            LOGGER.debug("HikariPool details: {}", details);

            if (!poolsAtLimit.isEmpty()) {
                LOGGER.warn("Hikari pools limit reached: {}", poolsAtLimit);
                threadDump();
            }
            return buildHealthResponse(!poolsAtLimit.isEmpty(), details);
        } catch (RuntimeException exception) {
            LOGGER.error("Failed while checking the health of hikari pool!", exception);
            return buildHealthResponse(false, Map.of());
        }
    }

Reply:
Thanks for the comment. It is more like a coding style preference. I believe the suggestion is less readable than the original one. Bit of overuse of the streaming API, which is harder to understand and harder to debug, and more error-prone.

Let's wait for other reviewers, and if comments form other reviews are suggesting that streaming API is easier to read in this case then I am happy to change it.

Reply:
I would vote for the old-school way due to the easier debugging because both of them are equally easy to read (to me at least)

Reply:
I would go with the original code because it is cleaner to read for me.